### PR TITLE
Optimization Controller for Review

### PR DESCRIPTION
**Pull request by AndrewHWang1605**
Configured with automated test (similar to pd_test), and passes.
Quite nice tracking behavior and gas expenditure on sim! Tested and runs on hardware (hardware experiments to test tracking/gas expenditure pending)

### CODE REVIEW

File: ff_control/ff_control/tri_thruster_ctrl.py [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    def __init__(self, node_name: str = "tri_thruster_ctrl_node") -> None:
+        super().__init__(node_name)
+
+    def set_tri_thrusters(self, tri_switches: T.Sequence[int], use_wheel: bool = False) -> None:
+        """
+        Convert trinary thruster commands into binary thruster commands
+        """
+        if use_wheel:
+            self.get_logger().error("set_tri_thrusters failed: use_wheel not implemented")
+            return
```
Comment (alvinsunyixiao) [inline ff_control/ff_control/tri_thruster_ctrl.py:40]: Let's remove that for now. I don't see this wheel getting implemented any time soon

File: ff_control/ff_control/tri_thruster_ctrl.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+
+class TrinaryThrusterController(LowLevelController):
+    def __init__(self, node_name: str = "tri_thruster_ctrl_node") -> None:
+        super().__init__(node_name)
+
+    def set_tri_thrusters(self, tri_switches: T.Sequence[int], use_wheel: bool = False) -> None:
+        """
+        Convert trinary thruster commands into binary thruster commands
+        """
```
Comment (alvinsunyixiao) [inline ff_control/ff_control/tri_thruster_ctrl.py:37]: Add some doc for `tri_switches`, e.g. explain what is the relationship / order with respect to the binary switches.

File: ff_control/ff_control/tri_thruster_ctrl.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+# MIT License
+#
+# Copyright (c) 2023 Stanford Autonomous Systems Lab
```
Comment (alvinsunyixiao) [inline ff_control/ff_control/tri_thruster_ctrl.py:3]: Change license to 2024. And same for all other new files.

File: ff_control/scripts/opt_ctrl_py_node [hunk 4]
```diff
@@ -0,0 +1,498 @@
...
+        kth = rclpy.Parameter('opt_gain_kth', rclpy.Parameter.Type.DOUBLE, 5.0)
+        kpos = rclpy.Parameter('opt_gain_kpos', rclpy.Parameter.Type.DOUBLE, 10.0)
+        kin = rclpy.Parameter('opt_gain_kinput', rclpy.Parameter.Type.DOUBLE, 3.5)
+        kvelo = rclpy.Parameter('opt_gain_kvelo', rclpy.Parameter.Type.DOUBLE, 20.5)
+
+        self.set_parameters([horizon, freq, kth, kpos, kin, kvelo])
+        self.w0, self.cont_nlp_solver, self.lbw, self.ubw, self.lbg, self.ubg = self.init_solver()
+
+        self.timer.cancel()
+        self.timer = self.create_timer(0.1, self.control_loop)
```
Comment (alvinsunyixiao) [inline ff_control/scripts/opt_ctrl_py_node:487]: Maybe it's worth considering having 2 sets of parameters. Switching params on the fly can be inefficient.

File: ff_sim/ff_sim/simulator_node.py [hunk 5]
```diff
@@ -116,6 +117,16 @@ def __init__(self):
...
+        self.running_total_gas = 0
+        self.thrust_hist = [[] for i in range(8)]
+        self.thrust_duty_cycles = [0] * 8
+        # self.start_time = self.get_clock().now().to_msg()
+        self.steps = 0
+        self.duty_cycle_window = 200
+        self.rolled_up = False
+        # self.get_logger().info("********MESSAGE"+str(self.start_time))
+        # self.get_logger().info("********SEC"+str(self.start_time.sec))
+
```
Comment (alvinsunyixiao) [inline ff_sim/ff_sim/simulator_node.py:129]: Remove?

File: ff_sim/ff_sim/simulator_node.py [hunk 6]
```diff
@@ -180,6 +191,9 @@ def __init__(self):
         self.declare_parameter("mocap_noise_xy", 0.001)
         self.declare_parameter("mocap_noise_theta", math.radians(0.1))
         self.pub_mocap = self.create_publisher(PoseStamped, "mocap/sim/pose", 10)
+        self.pub_controller_metrics = self.create_publisher(
+            ControllerMetrics, "controller/metrics", 10
+        )
```
Comment (alvinsunyixiao) [inline ff_sim/ff_sim/simulator_node.py:196]: remove

**Review by alvinsunyixiao — CHANGES_REQUESTED** [review]
Nice work overall! Left some comments on the details.

File: ff_control/ff_control/tri_thruster_ctrl.py (hunk 3 above)
Comment (AndrewHWang1605) [reply ff_control/ff_control/tri_thruster_ctrl.py:3]: For this license, do I only change the files that I worked on/changed? Or should I do this for anything my code interacts with?

File: ff_sim/ff_sim/controller_metrics.py [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+                self.rolled_up = True
+        else:
+            self.time_hist.pop(0)
+            self.time_hist.append(dt)
+            for i in range(8):
+                self.thrust_hist[i].pop(0)
+                self.thrust_hist[i].append(thrusters[i])
+                self.thrust_duty_cycles[i] = np.dot(self.thrust_hist[i], self.time_hist) / np.sum(
+                    self.time_hist
+                )
```
[the remark this answers is not in the document]
Comment (AndrewHWang1605) [reply ff_sim/ff_sim/controller_metrics.py:110]: Effectively, that's what this code is doing I believe, and I want it this way so the duty cycles are still valid even if we havent reached the full window yet

File: ff_control/scripts/opt_ctrl_py_node [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+
+import copy
+import rclpy
+
+import numpy as np
+from casadi import *
```
Comment (alvinsunyixiao) [inline ff_control/scripts/opt_ctrl_py_node:30]: Nit picking, but this is considered a bad style I believe

File: ff_control/scripts/opt_ctrl_py_node [hunk 9]
```diff
@@ -0,0 +1,505 @@
...
+        """
+        state = FreeFlyerState()
+        state.pose.x = vec[0]
+        state.pose.y = vec[1]
+        state.pose.theta = vec[2]
+        state.twist.vx = vec[3]
+        state.twist.vy = vec[4]
+        state.twist.wz = vec[5]
+
+        return state
```
Comment (alvinsunyixiao) [inline ff_control/scripts/opt_ctrl_py_node:395]: Is it copied from some other node? Maybe we can create a common utility function for this instead of duplicating code?

File: ff_control/scripts/opt_ctrl_py_node [hunk 10]
```diff
@@ -0,0 +1,505 @@
...
+
+
+    def set_close_hyperparameters(self) -> None:
+        # Optimization close-range fine-tuning trajectories hyperparameters/gains
+        self.get_logger().info("*** Set to close hyperparams ***")
+        self._def_controller = False
+        self.w0, self.cont_nlp_solver, self.lbw, self.ubw, self.lbg, self.ubg = self.init_solver()
+
+        self.timer.cancel()
+        self.timer = self.create_timer(1.0/self.freq, self.control_loop)
```
Comment (alvinsunyixiao) [inline ff_control/scripts/opt_ctrl_py_node:494]: Much cleaner than before!

File: ff_sim/ff_sim/controller_metrics.py [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+    1. total_gas_time: Measures total time that thrusters are on (summed over each thruster). Provides
+        a proxy for total gas expenditure over time
+    2. running_duty_cycles: Measures average duty cycle for each thruster over a time window specified by
+        self.duty_cycle_window.
+
+    """
+
+    def __init__(self):
+        super().__init__("ff_ctrl_metrics")
+        self.curr_time = self.get_clock().now().to_msg()
```
Comment (alvinsunyixiao) [inline ff_sim/ff_sim/controller_metrics.py:65]:
```suggestion
        self.curr_time = self.get_clock().now()
```

File: ff_sim/ff_sim/controller_metrics.py [hunk 12]
```diff
@@ -0,0 +1,139 @@
...
+    def process_new_wheel_cmd(self, msg: WheelVelCommand) -> None:
+        """Placeholder for now"""
+        pass
+
+    def process_new_binary_thrust_cmd(self, msg: ThrusterCommand) -> None:
+        """Process binary thrusters"""
+        now = self.get_clock().now().to_msg()
+        dtsec = now.sec - self.curr_time.sec
+        dtnsec = now.nanosec - self.curr_time.nanosec
+        dt = dtsec + dtnsec / 1e9
```
Comment (alvinsunyixiao) [inline ff_sim/ff_sim/controller_metrics.py:94]:
```suggestion
        now = self.get_clock().now()
        dt = (now - self.curr_time).nanoseconds / 1e9
```

File: ff_sim/ff_sim/controller_metrics.py [hunk 13]
```diff
@@ -0,0 +1,139 @@
...
+            self.time_hist.append(dt)
+            for i in range(8):
+                self.thrust_hist[i].append(thrusters[i])
+                self.thrust_duty_cycles[i] = np.dot(self.thrust_hist[i], self.time_hist) / np.sum(
+                    self.time_hist
+                )
+            if self.steps >= self.duty_cycle_window:
+                self.rolled_up = True
+        else:  # Once queue is filled up, can just treat this as a list that's constantly being updawted
+            self.time_hist.pop(0)
```
Comment (alvinsunyixiao) [inline ff_sim/ff_sim/controller_metrics.py:113]: Not sure how big will `self.time_hist` become, but `pop(0)` is an O(N) operation, so can be slow when the list becomes large. Suggest using an actual queue data structure instead of Python list.

File: ff_sim/ff_sim/simulator_node.py [hunk 14]
```diff
@@ -402,6 +402,13 @@ def compute_dynamics_dt(self, x_k, u_k, dt, discretization="Euler"):
         else:
             print("[FreeFlyerSimulator::compute_dynamics_dt]: Unknown Discretization Scheme.")
 
+        # Wrap theta to [-pi, pi]
+
+        # x_next[2] = ((x_next[2] % (2 * np.pi)) + 2 * np.pi) % (2 * np.pi)
+        # if x_next[2] > np.pi:
+        #     x_next[2] -= 2 * np.pi
```
Comment (alvinsunyixiao) [inline ff_sim/ff_sim/simulator_node.py:409]: remove?

File: ff_control/ff_control/tri_thruster_ctrl.py [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+        """
+        Convert trinary thruster commands into binary thruster commands
+        This formulation represents each thruster pair (eg thruster 1 and 2 below) as a single
+        "trinary" thruster, which can either take value -1 (1 on 2 off), 0 (both off), or 1 (1 off 2 on)
+        This reduces the search space for the optimization, and implicitly removes consideration of the
+        undesirable case where both thrusters are on (0 net force or moment, only wasted fuel)
+        tri_swtiches[0] = Thruster Pair [1,2]
+        tri_swtiches[1] = Thruster Pair [3,4]
+        tri_swtiches[2] = Thruster Pair [5,6]
+        tri_swtiches[3] = Thruster Pair [7,0]
```
Comment (alvinsunyixiao) [inline ff_control/ff_control/tri_thruster_ctrl.py:44]: Very nice!

File: ff_control/scripts/opt_ctrl_py_node [hunk 16]
```diff
@@ -0,0 +1,505 @@
...
+
+    ############################### Helper Functions for Opt ###############################
+    def _map_to_force(self, u):
+        # Compute body-frame force from thrusters
+        Fx = -u[0] + u[2] 
+        Fy = -u[1] + u[3] 
+        M = self.r * (u[0]+u[1]+u[2]+u[3])
+        return Fx, Fy, M
+
+    def normsq(self, x):
```
Comment (alvinsunyixiao) [inline ff_control/scripts/opt_ctrl_py_node:325]:
```suggestion
    def normsq(self, x):
        """ norm square operation for casadi variables. """
```

File: ff_control/scripts/opt_ctrl_py_node [hunk 17]
```diff
@@ -0,0 +1,505 @@
...
+        sum = 0
+        for i in range(x.shape[0]):
+            sum += x[i]**2
+        return sum
+
+    def normabs(self, x):
+        sum = 0
+        for i in range(x.shape[0]):
+            sum += fabs(x[i])
+        return sum
```
Comment (alvinsunyixiao) [inline ff_control/scripts/opt_ctrl_py_node:335]: Is this used anywhere?

File: testing_files/injectGoalPose.py
Comment (alvinsunyixiao) [inline testing_files/injectGoalPose.py:1]: suggest renaming the file to `inject_goal_pose.py` to make it consistent with other filenames.

File: testing_files/unpackRosBag.py
Comment (alvinsunyixiao) [inline testing_files/unpackRosBag.py:1]: Suggest renaming to `unpack_rosbag.py`. Also, is this the rosbag reader for the CoRL experiments? Looks like this might need some more documentation and cleanup. Maybe consider moving this to a separate PR?

**Review by alvinsunyixiao — COMMENTED** [review]
Awesome work, Andrew! This is very well written and documented. The code LGTM overall, left some comments mostly on the code styles and nitpicks.

File: testing_files/unpackRosBag.py
Comment (AndrewHWang1605) [reply testing_files/unpackRosBag.py:1]: This is a separate ROSBAG reader that I wrote myself to process data from my hardware experiments. Do you still think I should add it as a separate PR?

**Review by alvinsunyixiao — APPROVED** [review]
Awesome work!

**Comment by AndrewHWang1605** [comment]
The optimization-based controller implemented here is functional and experimentally verified, but continuing work to improve (primarily speed up the optimization past 3Hz default control frequency) is already under way, and will be incorporated in a future PR.
